### ISSUE TITLE
implementation detail - functional calls

### ISSUE THREAD

**Issue by AruniRC** [allenai/savn#9]
Hi, have an implementation related question:

In the `embedding()` method for the basemodel, there is an option for providing `params`. If params are provided, then the code uses functional Pytorch methods to call the forward passes of the layers: https://github.com/allenai/savn/blob/master/models/basemodel.py#L73

Was there any particular reason to use functional, and *not* simply load the params into the model state-dict at this phase? I am thinking of replacing the functional calls with a load_state_dict(), because custom layers may not always have a functional PyTorch implementation.... and wanted to make sure this would not break things (I will try it out in anycase...). 

Thanks again for your responses (I have been bugging you over most of this year! :D )

**Comment by mitchellnw** [comment]
Hi,

No worries at all, this is a great question!
We have the option to pass in `params` because, when using MAML-like second order methods methods in `pytorch`, everything needs to be done with `functional` methods. This is just an unfortunate consequence of how `pytorch` operates on model parameters -- and is currently an open issue on the `pytorch` github https://github.com/pytorch/pytorch/issues/12659.

However, the option to use the passed in params is only used in the meta-learning case (i.e. it is not used for the non-adaptive baseline). So if you don't need to do anything that requires a gradient through a gradient I wouldn't worry about this.
